[BUG] Dependencies not automatically resolved during build

### Current Behavior

Hi,

I keep on experiencing, during build that referenced assembly cannot be found. The csproj of the app has a ProjectReference to the project.

The workaround is that I have to add the referenced assembly as an implicit dependency to the app's project.json.

What could cause this? Could there any fragility in nx-dotnet's resolving of dependencies? As I understand it nx-dotnet should resolve probject dependencies automatically by inspecting csproj files?




### Expected Behavior

That nx-dotnet automatically resolves dependencies from csproj-files.

### Github Repo

_No response_

### Steps to Reproduce

1.


### Nx Report

```shell
Node   : 18.17.1
   OS     : darwin-arm64
   yarn   : 1.22.19
   
   nx                 : 17.2.8
   @nx/js             : 17.2.8
   @nx/jest           : 17.2.8
   @nx/linter         : 17.2.8
   @nx/eslint         : 17.2.8
   @nx/workspace      : 17.2.8
   @nx/cypress        : 17.2.8
   @nx/devkit         : 17.2.8
   @nx/eslint-plugin  : 17.2.8
   @nx/plugin         : 17.2.8
   @nx/react          : 17.2.8
   @nx/storybook      : 17.2.8
   @nrwl/tao          : 17.2.8
   @nx/web            : 17.2.8
   @nx/webpack        : 17.2.8
   typescript         : 5.2.2
   ---------------------------------------
   Community plugins:
   @nx-dotnet/core : 2.1.2
   ---------------------------------------
   Local workspace plugins:
         @airmont/nx-plugins

✨  Done in 1.66s.
```


### nx.json

```json
{
  "plugins": ["@nx-dotnet/core"],
  "namedInputs": {
    "default": ["{projectRoot}/**/*", "sharedGlobals"],
    "production": [
      "default",
      "!{projectRoot}/.eslintrc.json",
      "!{projectRoot}/**/?(*.)+(spec|test).[jt]s?(x)?(.snap)",
      "!{projectRoot}/tsconfig.spec.json",
      "!{projectRoot}/jest.config.[jt]s",
      "!{projectRoot}/.storybook/**/*",
      "!{projectRoot}/**/*.stories.@(js|jsx|ts|tsx|mdx)",
      "!{projectRoot}/src/test-setup.[jt]s",
      "!{projectRoot}/tsconfig.storybook.json"
    ],
    "sharedGlobals": ["{workspaceRoot}/babel.config.json"]
  },
  "targetDefaults": {
    "build": {
      "dependsOn": [
        "^build",
        "codegen",
        "^codegen",
        "codegen",
        "^codegen",
        "codegen",
        "^codegen"
      ],
      "inputs": ["production", "^production"],
      "cache": true
    },
    "e2e": {
      "inputs": ["default", "^production"],
      "cache": true
    },
    "lint": {
      "inputs": [
        "default",
        "{workspaceRoot}/.eslintrc.json",
        "{workspaceRoot}/.eslintignore"
      ],
      "cache": true
    },
    "test": {
      "inputs": ["default", "^production", "{workspaceRoot}/jest.preset.js"],
      "cache": true
    },
    "build-storybook": {
      "inputs": [
        "default",
        "^production",
        "{projectRoot}/.storybook/**/*",
        "{projectRoot}/tsconfig.storybook.json"
      ],
      "cache": true
    },
    "@nx/jest:jest": {
      "inputs": ["default", "^production", "{workspaceRoot}/jest.preset.js"],
      "cache": true,
      "options": {
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "generators": {
    "@nx/react": {
      "application": {
        "style": "styled-components",
        "linter": "eslint",
        "bundler": "webpack",
        "babel": true
      },
      "component": {
        "style": "styled-components"
      },
      "library": {
        "style": "styled-components",
        "linter": "eslint",
        "unitTestRunner": "jest"
      }
    }
  },
  "$schema": "./node_modules/nx/schemas/nx-schema.json"
}
```


### Failure Logs

```shell
> nx run firefly-my-chimney-webapi-cs:build:production

Executing Command: dotnet "--version"
8.0.101
Executing Command: dotnet "build" "C:\BuildAgent\_work\10\s\apps\firefly\my-chimney\webapi-cs\Airmont.Firefly.MyChimney.WebApiCs.csproj" "--configuration" "Release" "--no-dependencies" "false" "--verbosity" "minimal"
MSBuild version 17.8.3+195e7f5a3 for .NET
  Determining projects to restore...
C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Storage.Timeseries\Firefly.Storage.Timeseries.csproj : warning NU1902: Package 'System.Data.SqlClient' 4.8.4 has a known moderate severity vulnerability, https://github.com/advisories/GHSA-8g2p-5pqh-5jmc [C:\BuildAgent\_work\10\s\apps\firefly\my-chimney\webapi-cs\Airmont.Firefly.MyChimney.WebApiCs.csproj]
C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Storage.Timeseries\Firefly.Storage.Timeseries.csproj : warning NU1903: Package 'System.Data.SqlClient' 4.8.4 has a known high severity vulnerability, https://github.com/advisories/GHSA-98g6-xh36-x2p7 [C:\BuildAgent\_work\10\s\apps\firefly\my-chimney\webapi-cs\Airmont.Firefly.MyChimney.WebApiCs.csproj]
C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Provisioning.Services\Firefly.Provisioning.Services.csproj : warning NU1903: Package 'Azure.Identity' 1.7.0 has a known high severity vulnerability, https://github.com/advisories/GHSA-5mfx-4wcx-rv27 [C:\BuildAgent\_work\10\s\apps\firefly\my-chimney\webapi-cs\Airmont.Firefly.MyChimney.WebApiCs.csproj]
C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Storage.KeyVault\Firefly.Storage.KeyVault.csproj : warning NU1903: Package 'Azure.Identity' 1.7.0 has a known high severity vulnerability, https://github.com/advisories/GHSA-5mfx-4wcx-rv27 [C:\BuildAgent\_work\10\s\apps\firefly\my-chimney\webapi-cs\Airmont.Firefly.MyChimney.WebApiCs.csproj]
  Restored C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Provisioning.Data\Firefly.Provisioning.Data.csproj (in 451 ms).
  Restored C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Storage.KeyVault\Firefly.Storage.KeyVault.csproj (in 451 ms).
  Restored C:\BuildAgent\_work\10\s\libs\firefly\my-chimney\cs\service\service.csproj (in 531 ms).
  Restored C:\BuildAgent\_work\10\s\libs\firefly\shared\cs\Firefly.Shared.Service\Firefly.Shared.Service.csproj (in 530 ms).
  Restored C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Core.Application.Web\Firefly.Core.Application.Web.csproj (in 531 ms).
  Restored C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Core.Web\Firefly.Core.Web.csproj (in 468 ms).
  Restored C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Provisioning.Services\Firefly.Provisioning.Services.csproj (in 531 ms).
  Restored C:\BuildAgent\_work\10\s\apps\firefly\my-chimney\webapi-cs\Airmont.Firefly.MyChimney.WebApiCs.csproj (in 535 ms).
  9 of 17 projects are up-to-date for restore.
C:\Program Files\dotnet\sdk\8.0.101\Sdks\Microsoft.NET.Sdk.Razor\targets\Microsoft.NET.Sdk.Razor.MvcApplicationPartsDiscovery.targets(54,5): warning RAZORSDK1007: Reference assembly C:\BuildAgent\_work\10\s\dist\libs\Firefly\shared\cs\Firefly.Core.Application.Web\net6.0\Firefly.Core.Application.Web.dll could not be found. This is typically caused by build errors in referenced projects. [C:\BuildAgent\_work\10\s\apps\firefly\my-chimney\webapi-cs\Airmont.Firefly.MyChimney.WebApiCs.csproj]
CSC : error CS0006: Metadata file 'C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Core.Application.Web\obj\Release\net6.0\ref\Firefly.Core.Application.Web.dll' could not be found [C:\BuildAgent\_work\10\s\apps\firefly\my-chimney\webapi-cs\Airmont.Firefly.MyChimney.WebApiCs.csproj]
CSC : error CS0006: Metadata file 'C:\BuildAgent\_work\10\s\libs\Firefly\shared\cs\Firefly.Storage.KeyVault\obj\Release\net6.0\ref\Firefly.Storage.KeyVault.dll' could not be found [C:\BuildAgent\_work\10\s\apps\firefly\my-chimney\webapi-cs\Airmont.Firefly.MyChimney.WebApiCs.csproj]

Build FAILED.
```


### Additional Information

csproj of the app (firefly-my-chimney-webapi-cs) that fails during build:
```
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
    <UserSecretsId>7813e2b2-f546-44d9-bf13-ca8b6ecf45df</UserSecretsId>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.2.3" />
    <PackageReference Include="Microsoft.AspNetCore.SpaServices.Extensions" Version="6.0.7" />
    <PackageReference Include="Serilog.AspNetCore" Version="6.0.1" />
    <PackageReference Include="Serilog.Enrichers.Environment" Version="2.2.0" />
    <PackageReference Include="Serilog.Sinks.ApplicationInsights" Version="4.0.0" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\..\..\..\libs\firefly\my-chimney\cs\service\service.csproj" />
    <ProjectReference Include="..\..\..\..\libs\firefly\my-chimney\cs\Ef\Ef.csproj" />
    <ProjectReference Include="..\..\..\..\libs\Firefly\shared\cs\Firefly.Core.Application.Web\Firefly.Core.Application.Web.csproj" />
    <ProjectReference Include="..\..\..\..\libs\firefly\my-chimney\cs\domain\domain.csproj"/>
    <ProjectReference Include="..\..\..\..\libs\Firefly\shared\cs\Firefly.Provisioning.Services\Firefly.Provisioning.Services.csproj" />
    <ProjectReference Include="..\..\..\..\libs\Firefly\shared\cs\Firefly.Storage.DocumentDb\Firefly.Storage.DocumentDb.csproj" />
    <ProjectReference Include="..\..\..\..\libs\Firefly\shared\cs\Firefly.Storage.KeyVault\Firefly.Storage.KeyVault.csproj" />
  </ItemGroup>

  <ItemGroup>
    <Content Remove="project.json" />
    <None Include="project.json">
      <CopyToOutputDirectory>Never</CopyToOutputDirectory>
    </None>
  </ItemGroup>


</Project>
```

Same issue. In my opinion it's something wrong with "noDependencies".

When I run `nx build <project_name>` with below `project.json` configuration:
```
  "targets": {
    "build": {
      "options": {
        "noDependencies": false
      }
    }
  }
 ```
 Everything works as indented. This runs: `dotnet "build" "/home/x/y/z.csproj" "--configuration" "Debug" "--verbosity" "minimal"`
 
 But when I run same command with this configuration (default one):
 ```
  "targets": {
    "build": {
      "options": {
        "noDependencies": true
      }
    }
  }
 ```
 I get error `CSC : error CS0006: Metadata file`: This runs: `dotnet "build" "/home/x/y/z.csproj" "--configuration" "Debug" "--no-dependencies" "--verbosity" "minimal"`.
 
 There is something wrong with: https://www.nx-dotnet.com/docs/core/guides/incremental-builds

I found issue. MSBuild properties are not supported for `nx`.

Example:
If in `.csproj` i use relative path:
```
(...) 
<ProjectReference Include="..\..\..\..\libs\x\x.csproj" />
```
Build with "--no-dependencies" works and `nx graph` show everything as expected.

But when i use MSBuild properties like this:
`Directory.Build.props`:
```
    <RepoRoot>$([System.IO.Path]::GetFullPath('$(MSBuildThisFileDirectory)'))</RepoRoot>
```
`.csproj`:
```
(...) 
<ProjectReference Include="$(RepoRoot)\libs\x\x.csproj" />
```

`nx graph` show no connection with my `x` library and build with "--no-dependencies" throw `CSC : error CS0006: Metadata file`

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -12,7 +12,7 @@ concurrency:
 jobs:
   main:
     name: Nx Cloud - Main Job
-    uses: nrwl/ci/.github/workflows/nx-cloud-main.yml@v0.8
+    uses: nrwl/ci/.github/workflows/nx-cloud-main.yml@v0.15
     with:
       main-branch-name: 'master'
       parallel-commands: |
@@ -25,6 +25,6 @@ jobs:
 
   agents:
     name: Nx Cloud - Agents
-    uses: nrwl/ci/.github/workflows/nx-cloud-agents.yml@v0.8
+    uses: nrwl/ci/.github/workflows/nx-cloud-agents.yml@v0.15
     with:
       number-of-agents: 3

--- a/.github/workflows/pr.yml
+++ b/.github/workflows/pr.yml
@@ -14,7 +14,7 @@ concurrency:
 jobs:
   main:
     name: Nx Cloud - Main Job
-    uses: nrwl/ci/.github/workflows/nx-cloud-main.yml@v0.8
+    uses: nrwl/ci/.github/workflows/nx-cloud-main.yml@v0.15
     with:
       main-branch-name: 'master'
       parallel-commands: |
@@ -26,6 +26,6 @@ jobs:
 
   agents:
     name: Nx Cloud - Agents
-    uses: nrwl/ci/.github/workflows/nx-cloud-agents.yml@v0.8
+    uses: nrwl/ci/.github/workflows/nx-cloud-agents.yml@v0.15
     with:
       number-of-agents: 3

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -639,7 +639,7 @@ Co-authored-by: Craigory Coppola <craigorycoppola@gmail.com>
 
 ### Bug Fixes
 
-- **utils:** getDependantProjectsForNxProject should work on Unix ([96cbc33](https://github.com/nx-dotnet/nx-dotnet/commit/96cbc33ec6b5e9d0492fba4902ee76938230b146)), closes [#43](https://github.com/nx-dotnet/nx-dotnet/issues/43)
+- **utils:** forEachDependantProject should work on Unix ([96cbc33](https://github.com/nx-dotnet/nx-dotnet/commit/96cbc33ec6b5e9d0492fba4902ee76938230b146)), closes [#43](https://github.com/nx-dotnet/nx-dotnet/issues/43)
 
 ## [0.6.1](https://github.com/nx-dotnet/nx-dotnet/compare/v0.6.0...v0.6.1) (2021-05-19)
 

--- a/nx.json
+++ b/nx.json
@@ -8,7 +8,8 @@
         "nugetPackages": {
           "Swashbuckle.AspNetCore": "6.5.0"
         }
-      }
+      },
+      "include": ["demo/**/*"]
     }
   ],
   "generators": {

--- a/packages/core/src/graph/create-dependencies.spec.ts
+++ b/packages/core/src/graph/create-dependencies.spec.ts
@@ -1,0 +1,22 @@
+import { workspaceRoot } from '@nx/devkit';
+import { resolveReferenceToProject } from './create-dependencies';
+
+describe('createDependencies', () => {
+  describe('resolveReferenceToProject', () => {
+    it('should find project in rootMap', () => {
+      expect(
+        resolveReferenceToProject(
+          '../../libs/my-lib/MyLib.csproj',
+          'apps/my-app/MyApp.csproj',
+          {
+            'libs/my-lib': 'my-lib',
+            'apps/my-app': 'my-app',
+          },
+          {
+            workspaceRoot,
+          },
+        ),
+      ).toEqual('my-lib');
+    });
+  });
+});

--- a/packages/core/src/graph/create-dependencies.ts
+++ b/packages/core/src/graph/create-dependencies.ts
@@ -1,18 +1,18 @@
 import {
   CreateDependencies,
   CreateDependenciesContext,
+  DependencyType,
   NxPluginV1,
+  ProjectConfiguration,
   ProjectGraphBuilder,
   RawProjectGraphDependency,
+  normalizePath,
   workspaceRoot,
 } from '@nx/devkit';
-import { parse } from 'node:path';
+import { dirname, parse, relative, resolve } from 'node:path';
 
-import {
-  getDependenciesFromXmlFile,
-  NxDotnetConfig,
-  readConfig,
-} from '@nx-dotnet/utils';
+import { NxDotnetConfig, readConfig } from '@nx-dotnet/utils';
+import { DotNetClient, dotnetFactory } from '@nx-dotnet/dotnet';
 
 // Between Nx versions 16.8 and 17, the signature of `CreateDependencies` changed.
 // It used to only consist of the context, but now it also includes the options.
@@ -29,6 +29,8 @@ type CreateDependenciesCompat<T> = {
   ): ReturnType<CreateDependencies<T>>;
 };
 
+const dotnetClient = new DotNetClient(dotnetFactory());
+
 export const createDependencies: CreateDependenciesCompat<NxDotnetConfig> = (
   ctxOrOpts: CreateDependenciesContext | NxDotnetConfig | undefined,
   maybeCtx: CreateDependenciesContext | undefined,
@@ -39,17 +41,35 @@ export const createDependencies: CreateDependenciesCompat<NxDotnetConfig> = (
     maybeCtx ?? (ctxOrOpts as CreateDependenciesContext);
 
   let dependencies: RawProjectGraphDependency[] = [];
-  const rootMap = Object.fromEntries(
-    Object.entries(ctx.projects).map(([name, project]) => [project.root, name]),
-  );
+  const rootMap = createProjectRootMappings(ctx.projects);
   for (const source in ctx.filesToProcess.projectFileMap) {
     const changed = ctx.filesToProcess.projectFileMap[source];
     for (const file of changed) {
       const { ext } = parse(file.file);
       if (['.csproj', '.fsproj', '.vbproj'].includes(ext)) {
-        dependencies = dependencies.concat(
-          getDependenciesFromXmlFile(file.file, source, rootMap),
-        );
+        const references = dotnetClient.getProjectReferences(file.file);
+        const newDeps: RawProjectGraphDependency[] = [];
+        for (const reference of references) {
+          const project = resolveReferenceToProject(
+            normalizePath(reference),
+            file.file,
+            rootMap,
+            ctx,
+          );
+          if (project) {
+            newDeps.push({
+              source,
+              target: project,
+              type: DependencyType.static,
+              sourceFile: file.file,
+            });
+          } else {
+            console.warn(
+              `Unable to resolve project for reference ${reference} in ${file.file}`,
+            );
+          }
+        }
+        dependencies = dependencies.concat(newDeps);
       }
     }
   }
@@ -78,3 +98,48 @@ export const processProjectGraph: Required<NxPluginV1>['processProjectGraph'] =
     }
     return builder.getUpdatedProjectGraph();
   };
+
+function createProjectRootMappings(
+  projects: Record<string, ProjectConfiguration>,
+) {
+  const rootMap: Record<string, string> = {};
+  for (const [, project] of Object.entries(projects)) {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    rootMap[project.root!] = project.name!;
+  }
+  return rootMap;
+}
+
+function findProjectForPath(
+  filePath: string,
+  rootMap: Record<string, string>,
+): string | undefined {
+  /**
+   * Project Mappings are in UNIX-style file paths
+   * Windows may pass Win-style file paths
+   * Ensure filePath is in UNIX-style
+   */
+  let currentPath = normalizePath(filePath);
+  for (
+    ;
+    currentPath !== dirname(currentPath);
+    currentPath = dirname(currentPath)
+  ) {
+    const p = rootMap[currentPath];
+    if (p) {
+      return p;
+    }
+  }
+  return rootMap[currentPath];
+}
+
+export function resolveReferenceToProject(
+  reference: string,
+  source: string,
+  rootMap: Record<string, string>,
+  context: { workspaceRoot: string },
+) {
+  const resolved = resolve(context.workspaceRoot, dirname(source), reference);
+  console.log({ reference, source, resolved });
+  return findProjectForPath(relative(context.workspaceRoot, resolved), rootMap);
+}

--- a/packages/core/src/tasks/check-module-boundaries.spec.ts
+++ b/packages/core/src/tasks/check-module-boundaries.spec.ts
@@ -131,11 +131,19 @@ describe('enforce-module-boundaries', () => {
   it('should exit early if no tags on project', async () => {
     const spy = jest.spyOn(checkModule, 'loadModuleBoundaries');
     const results = await checkModuleBoundariesForProject('a', {
-      a: {
-        tags: [],
-        targets: {},
-        root: '',
+      nodes: {
+        a: {
+          type: 'app',
+          name: 'a',
+          data: {
+            tags: [],
+            targets: {},
+            root: '',
+          },
+        },
       },
+      externalNodes: {},
+      dependencies: {},
     });
     expect(spy).not.toHaveBeenCalled();
     expect(results).toHaveLength(0);
@@ -151,15 +159,35 @@ describe('enforce-module-boundaries', () => {
     });
 
     const results = await checkModuleBoundariesForProject('a', {
-      a: {
-        tags: ['a'],
-        targets: { ui: {} },
-        root: 'libs/a',
+      externalNodes: {},
+      dependencies: {
+        a: [
+          {
+            source: 'a',
+            target: 'ui',
+            type: 'static',
+          },
+        ],
       },
-      ui: {
-        tags: ['ui'],
-        targets: {},
-        root: 'libs/ui',
+      nodes: {
+        a: {
+          type: 'app',
+          name: 'a',
+          data: {
+            tags: ['a'],
+            targets: { ui: {} },
+            root: 'libs/a',
+          },
+        },
+        ui: {
+          type: 'lib',
+          name: 'ui',
+          data: {
+            tags: ['ui'],
+            targets: {},
+            root: 'libs/ui',
+          },
+        },
       },
     });
     expect(results).toHaveLength(1);
@@ -175,15 +203,35 @@ describe('enforce-module-boundaries', () => {
     });
 
     const results = await checkModuleBoundariesForProject('b', {
-      a: {
-        tags: ['a'],
-        targets: {},
-        root: 'libs/a',
+      externalNodes: {},
+      dependencies: {
+        b: [
+          {
+            source: 'b',
+            target: 'a',
+            type: 'static',
+          },
+        ],
       },
-      b: {
-        tags: ['b'],
-        targets: { a: {} },
-        root: 'libs/b',
+      nodes: {
+        a: {
+          type: 'lib',
+          name: 'a',
+          data: {
+            tags: ['a'],
+            targets: {},
+            root: 'libs/a',
+          },
+        },
+        b: {
+          type: 'lib',
+          name: 'b',
+          data: {
+            tags: ['b'],
+            targets: { a: {} },
+            root: 'libs/b',
+          },
+        },
       },
     });
     expect(results).toHaveLength(1);
@@ -199,15 +247,27 @@ describe('enforce-module-boundaries', () => {
     });
 
     const results = await checkModuleBoundariesForProject('a', {
-      a: {
-        tags: ['a'],
-        targets: { shared: {} },
-        root: 'libs/a',
-      },
-      shared: {
-        tags: ['shared'],
-        targets: {},
-        root: 'libs/shared',
+      externalNodes: {},
+      dependencies: {},
+      nodes: {
+        a: {
+          type: 'app',
+          name: 'a',
+          data: {
+            tags: ['a'],
+            targets: { shared: {} },
+            root: 'libs/a',
+          },
+        },
+        shared: {
+          type: 'lib',
+          name: 'shared',
+          data: {
+            tags: ['shared'],
+            targets: {},
+            root: 'libs/shared',
+          },
+        },
       },
     });
     expect(results).toHaveLength(0);
@@ -223,39 +283,75 @@ describe('enforce-module-boundaries', () => {
     });
 
     const results = await checkModuleBoundariesForProject('x', {
-      x: {
-        tags: ['x', 'not-z'],
-        targets: { z: {} },
-        root: 'libs/x',
+      externalNodes: {},
+      dependencies: {
+        x: [
+          {
+            source: 'x',
+            target: 'z',
+            type: 'static',
+          },
+        ],
       },
-      z: {
-        tags: ['z'],
-        targets: {},
-        root: 'libs/z',
+      nodes: {
+        x: {
+          type: 'lib',
+          name: 'x',
+          data: {
+            tags: ['x', 'not-z'],
+            targets: { z: {} },
+            root: 'libs/x',
+          },
+        },
+        z: {
+          type: 'lib',
+          name: 'z',
+          data: {
+            tags: ['z'],
+            targets: {},
+            root: 'libs/z',
+          },
+        },
       },
     });
     expect(results).toHaveLength(1);
   });
 
   it('should find violations with allSourceTags/onlyDependOnLibsWithTags', async () => {
-    const globResults = ['libs/x/x.csproj'];
-    jest.spyOn(fastGlob, 'sync').mockImplementation(() => globResults);
-
     vol.fromJSON({
       'libs/x/x.csproj':
-        '<Project Sdk="Microsoft.NET.Sdk.Web"><ItemGroup><ProjectReference Include="..\\..\\libs\\a\\a.csproj" /></ItemGroup></Project>',
+        '<Project Sdk="Microsoft.NET.Sdk.Web"><ItemGroup><ProjectReference Include="..\\..\\libs\\z\\z.csproj" /></ItemGroup></Project>',
     });
-
     const results = await checkModuleBoundariesForProject('x', {
-      x: {
-        tags: ['x', 'only-z'],
-        targets: { a: {} },
-        root: 'libs/x',
+      externalNodes: {},
+      dependencies: {
+        x: [
+          {
+            source: 'x',
+            target: 'a',
+            type: 'static',
+          },
+        ],
       },
-      a: {
-        tags: ['a'],
-        targets: {},
-        root: 'libs/a',
+      nodes: {
+        x: {
+          type: 'lib',
+          name: 'x',
+          data: {
+            tags: ['x', 'only-z'],
+            targets: { a: {} },
+            root: 'libs/x',
+          },
+        },
+        a: {
+          type: 'lib',
+          name: 'a',
+          data: {
+            tags: ['a'],
+            targets: {},
+            root: 'libs/a',
+          },
+        },
       },
     });
     expect(results).toHaveLength(1);
@@ -271,15 +367,35 @@ describe('enforce-module-boundaries', () => {
     });
 
     const results = await checkModuleBoundariesForProject('x', {
-      x: {
-        tags: ['x', 'only-z'],
-        targets: { z: {} },
-        root: 'libs/x',
+      externalNodes: {},
+      dependencies: {
+        x: [
+          {
+            source: 'x',
+            target: 'z',
+            type: 'static',
+          },
+        ],
       },
-      z: {
-        tags: ['z'],
-        targets: {},
-        root: 'libs/z',
+      nodes: {
+        x: {
+          type: 'lib',
+          name: 'x',
+          data: {
+            tags: ['x', 'only-z'],
+            targets: { z: {} },
+            root: 'libs/x',
+          },
+        },
+        z: {
+          type: 'lib',
+          name: 'z',
+          data: {
+            tags: ['z'],
+            targets: {},
+            root: 'libs/z',
+          },
+        },
       },
     });
     expect(results).toHaveLength(0);
@@ -295,15 +411,35 @@ describe('enforce-module-boundaries', () => {
     });
 
     const results = await checkModuleBoundariesForProject('x', {
-      x: {
-        tags: ['no-deps'],
-        targets: { a: {} },
-        root: 'libs/x',
+      externalNodes: {},
+      dependencies: {
+        x: [
+          {
+            source: 'x',
+            target: 'a',
+            type: 'static',
+          },
+        ],
       },
-      a: {
-        tags: ['a'],
-        targets: {},
-        root: 'libs/a',
+      nodes: {
+        x: {
+          type: 'lib',
+          name: 'x',
+          data: {
+            tags: ['no-deps'],
+            targets: { a: {} },
+            root: 'libs/x',
+          },
+        },
+        a: {
+          type: 'lib',
+          name: 'a',
+          data: {
+            tags: ['a'],
+            targets: {},
+            root: 'libs/a',
+          },
+        },
       },
     });
     expect(results).toHaveLength(1);
@@ -319,15 +455,35 @@ describe('enforce-module-boundaries', () => {
     });
 
     const results = await checkModuleBoundariesForProject('x', {
-      x: {
-        tags: ['--foo--'],
-        targets: { a: {} },
-        root: 'libs/x',
+      externalNodes: {},
+      dependencies: {
+        x: [
+          {
+            source: 'x',
+            target: 'a',
+            type: 'static',
+          },
+        ],
       },
-      a: {
-        tags: ['biz'],
-        targets: {},
-        root: 'libs/a',
+      nodes: {
+        x: {
+          type: 'lib',
+          name: 'x',
+          data: {
+            tags: ['--foo--'],
+            targets: { a: {} },
+            root: 'libs/x',
+          },
+        },
+        a: {
+          type: 'lib',
+          name: 'a',
+          data: {
+            tags: ['biz'],
+            targets: {},
+            root: 'libs/a',
+          },
+        },
       },
     });
     expect(results).toHaveLength(2);
@@ -343,15 +499,35 @@ describe('enforce-module-boundaries', () => {
     });
 
     const results = await checkModuleBoundariesForProject('x', {
-      x: {
-        tags: ['==foo=='],
-        targets: { a: {} },
-        root: 'libs/x',
+      externalNodes: {},
+      dependencies: {
+        x: [
+          {
+            source: 'x',
+            target: 'a',
+            type: 'static',
+          },
+        ],
       },
-      a: {
-        tags: ['biz'],
-        targets: {},
-        root: 'libs/a',
+      nodes: {
+        x: {
+          type: 'lib',
+          name: 'x',
+          data: {
+            tags: ['==foo=='],
+            targets: { a: {} },
+            root: 'libs/x',
+          },
+        },
+        a: {
+          type: 'lib',
+          name: 'a',
+          data: {
+            tags: ['biz'],
+            targets: {},
+            root: 'libs/a',
+          },
+        },
       },
     });
     expect(results).toHaveLength(2);

--- a/packages/core/src/tasks/check-module-boundaries.ts
+++ b/packages/core/src/tasks/check-module-boundaries.ts
@@ -1,8 +1,7 @@
 import {
   createProjectGraphAsync,
-  ProjectConfiguration,
-  ProjectsConfigurations,
-  readProjectsConfigurationFromProjectGraph,
+  ProjectGraph,
+  ProjectGraphProjectNode,
   Tree,
   workspaceRoot,
 } from '@nx/devkit';
@@ -10,7 +9,7 @@ import {
 import { relative } from 'path';
 
 import {
-  getDependantProjectsForNxProject,
+  forEachDependantProject,
   ModuleBoundaries,
   ModuleBoundary,
   readConfig,
@@ -18,10 +17,10 @@ import {
 
 export async function checkModuleBoundariesForProject(
   project: string,
-  projects: Record<string, ProjectConfiguration>,
+  graph: ProjectGraph,
 ): Promise<string[]> {
-  const projectRoot = projects[project].root;
-  const tags = projects[project].tags ?? [];
+  const projectRoot = graph.nodes[project].data.root;
+  const tags = graph.nodes[project].data.tags ?? [];
   if (!tags.length) {
     return [];
   }
@@ -31,23 +30,19 @@ export async function checkModuleBoundariesForProject(
   }
 
   const violations: string[] = [];
-  getDependantProjectsForNxProject(
-    project,
-    { version: 2, projects },
-    (configuration, name, implicit) => {
-      if (implicit) return;
-      const dependencyTags = configuration?.tags ?? [];
-      for (const constraint of constraints) {
-        if (hasConstraintViolation(constraint, dependencyTags)) {
-          violations.push(
-            `${project} cannot depend on ${name}. Project tag ${JSON.stringify(
-              constraint,
-            )} is not satisfied.`,
-          );
-        }
+  forEachDependantProject(project, graph, (configuration, name, implicit) => {
+    if (implicit) return;
+    const dependencyTags = configuration?.tags ?? [];
+    for (const constraint of constraints) {
+      if (hasConstraintViolation(constraint, dependencyTags)) {
+        violations.push(
+          `${project} cannot depend on ${name}. Project tag ${JSON.stringify(
+            constraint,
+          )} is not satisfied.`,
+        );
       }
-    },
-  );
+    }
+  });
   return violations;
 }
 
@@ -101,7 +96,7 @@ export async function loadModuleBoundaries(
         result.rules['@nrwl/nx/enforce-module-boundaries'] ||
         [];
       return moduleBoundaryConfig?.depConstraints ?? [];
-    } catch {
+    } catch (e) {
       return [];
     }
   } else {
@@ -109,18 +104,18 @@ export async function loadModuleBoundaries(
   }
 }
 
-function findProjectGivenRoot(
-  root: string,
-  projects: ProjectsConfigurations['projects'],
-): string {
+function findProjectGivenRoot(root: string, graph: ProjectGraph): string {
   // Note that this returns the first matching project and would succeed for multiple (cs|fs...)proj under an nx project path,
   // but getProjectFileForNxProject explicitly throws if it's not exactly one.
   const normalizedRoot = root.replace(/^["'](.+(?=["']$))["']$/, '$1');
-  const [projectName] =
-    Object.entries(projects).find(([, projectConfig]) => {
-      const relativePath = relative(projectConfig.root, normalizedRoot);
+  const projectNode = Object.values(graph.nodes).find(
+    (projectConfig: ProjectGraphProjectNode) => {
+      const relativePath = relative(projectConfig.data.root, normalizedRoot);
       return relativePath?.startsWith('..') === false;
-    }) ?? [];
+    },
+  );
+
+  const projectName = projectNode?.name;
 
   if (projectName) {
     return projectName;
@@ -176,15 +171,13 @@ async function main() {
     string: ['project', 'projectRoot'],
   }) as { project?: string; projectRoot?: string };
   const graph = await createProjectGraphAsync();
-  const { projects }: ProjectsConfigurations =
-    readProjectsConfigurationFromProjectGraph(graph);
 
   // Find the associated nx project for the msbuild project directory.
   const nxProject: string =
-    project ?? findProjectGivenRoot(projectRoot as string, projects);
+    project ?? findProjectGivenRoot(projectRoot as string, graph);
 
   console.log(`Checking module boundaries for ${nxProject}`);
-  const violations = await checkModuleBoundariesForProject(nxProject, projects);
+  const violations = await checkModuleBoundariesForProject(nxProject, graph);
   if (violations.length) {
     violations.forEach((error) => {
       console.error(error);

--- a/packages/dotnet/src/lib/core/dotnet.client.ts
+++ b/packages/dotnet/src/lib/core/dotnet.client.ts
@@ -311,6 +311,23 @@ export class DotNetClient {
     this.logAndExecute(params);
   }
 
+  getProjectReferences(projectFile: string): string[] {
+    const output = this.spawnAndGetOutput(['list', projectFile, 'reference']);
+    // Output looks like:
+    // ```
+    // HEADER
+    // -------------------
+    // A
+    // B
+    // C
+    // ```
+    return output
+      .split('\n')
+      .slice(2)
+      .map((line) => line.trim())
+      .filter(Boolean);
+  }
+
   getSdkVersion(): string {
     return this.cliCommand.info.version.toString();
   }

--- a/packages/utils/fixtures/cs/file.csproj
+++ b/packages/utils/fixtures/cs/file.csproj
@@ -1,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <OutputType>Exe</OutputType>
+        <TargetFramework>net5.0</TargetFramework>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
+    </ItemGroup>
+
+</Project>

--- a/packages/utils/fixtures/fs/file.fsproj
+++ b/packages/utils/fixtures/fs/file.fsproj
@@ -1,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net5.0</TargetFramework>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <!-- Add your project files here -->
+    </ItemGroup>
+
+</Project>

--- a/packages/utils/fixtures/vb/file.vbproj
+++ b/packages/utils/fixtures/vb/file.vbproj
@@ -1,0 +1,6 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+    <PropertyGroup>
+        <OutputType>Library</OutputType>
+        <TargetFramework>net5.0</TargetFramework>
+    </PropertyGroup>
+</Project>

--- a/packages/utils/src/lib/utility-functions/workspace.ts
+++ b/packages/utils/src/lib/utility-functions/workspace.ts
@@ -4,22 +4,19 @@ import {
   normalizePath as nxNormalizePath,
   NX_VERSION,
   ProjectConfiguration,
-  ProjectsConfigurations,
-  RawProjectGraphDependency,
   Tree,
-  workspaceRoot,
   readNxJson,
   readJsonFile,
   NxJsonConfiguration,
+  ProjectGraph,
+  reverse,
 } from '@nx/devkit';
 
 import { readFileSync } from 'fs';
-import { dirname, relative, resolve } from 'path';
 import { lt } from 'semver';
-import { XmlDocument, XmlElement } from 'xmldoc';
+import { XmlDocument } from 'xmldoc';
 
 import { findProjectFileInPath, findProjectFileInPathSync, glob } from './glob';
-import { getAbsolutePath } from './path';
 
 export async function getProjectFileForNxProject(
   project: ProjectConfiguration,
@@ -34,111 +31,24 @@ export function getProjectFileForNxProjectSync(project: ProjectConfiguration) {
   return findProjectFileInPathSync(srcDirectory);
 }
 
-export function getDependantProjectsForNxProject(
+export function forEachDependantProject(
   targetProject: string,
-  projectsConfiguration: ProjectsConfigurations,
+  graph: ProjectGraph,
   forEachCallback?: (
     project: ProjectConfiguration,
     projectName: string,
     implicit: boolean,
   ) => void,
-): {
-  [projectName: string]: ProjectConfiguration;
-} {
-  const projectRoots: { [key: string]: string } = {};
-  const dependantProjects: { [key: string]: ProjectConfiguration } = {};
-
-  Object.entries(projectsConfiguration.projects).forEach(([name, project]) => {
-    projectRoots[project.root] = name;
-  });
-
-  const absoluteNetProjectFilePath = resolve(
-    workspaceRoot,
-    getProjectFileForNxProjectSync(
-      projectsConfiguration.projects[targetProject],
-    ),
-  );
-
-  const dependencies = getDependenciesFromXmlFile(
-    absoluteNetProjectFilePath,
-    targetProject,
-    projectRoots,
-  );
-
-  for (const dependency of dependencies) {
-    const project = projectsConfiguration.projects[dependency.target];
-    if (!project) {
-      throw new Error(
-        `Unable to find project ${dependency.target} in workspace`,
-      );
-    }
+) {
+  for (const dependant of graph.dependencies[targetProject] ?? []) {
     if (forEachCallback) {
       forEachCallback(
-        project,
-        dependency.target,
-        dependency.type === DependencyType.implicit,
+        graph.nodes[dependant.target].data,
+        dependant.target,
+        dependant.type === DependencyType.implicit,
       );
     }
-    dependantProjects[dependency.target] = project;
   }
-
-  return dependantProjects;
-}
-
-export function getDependenciesFromXmlFile(
-  filePath: string,
-  source: string,
-  projectRootMap: Record<string, string>,
-): RawProjectGraphDependency[] {
-  const found: RawProjectGraphDependency[] = [];
-
-  const absoluteNetProjectFilePath = getAbsolutePath(filePath, workspaceRoot);
-  const netProjectFilePath = normalizePath(
-    relative(workspaceRoot, absoluteNetProjectFilePath),
-  );
-  const hostProjectDirectory = normalizePath(
-    dirname(absoluteNetProjectFilePath),
-  );
-
-  const xml: XmlDocument | null = tryGetXmlDocument(absoluteNetProjectFilePath);
-
-  if (!xml) {
-    return found;
-  }
-
-  xml.childrenNamed('ItemGroup').forEach((itemGroup) =>
-    itemGroup.childrenNamed('ProjectReference').forEach((x: XmlElement) => {
-      const includeFilePath = normalizePath(x.attr['Include']);
-      const implicit = x.attr['ReferenceOutputAssembly'] === 'false';
-      const absoluteIncludedPath = getAbsolutePath(
-        includeFilePath,
-        hostProjectDirectory,
-      );
-      const workspaceFilePath = normalizePath(
-        relative(workspaceRoot, absoluteIncludedPath),
-      );
-
-      let potentialTargetRoot = dirname(workspaceFilePath);
-      while (
-        potentialTargetRoot !== workspaceRoot &&
-        potentialTargetRoot !== '.'
-      ) {
-        if (projectRootMap[potentialTargetRoot]) {
-          found.push({
-            source,
-            target: projectRootMap[potentialTargetRoot],
-            type: implicit ? DependencyType.implicit : DependencyType.static,
-            sourceFile: netProjectFilePath,
-          });
-          break;
-        }
-
-        potentialTargetRoot = dirname(potentialTargetRoot);
-      }
-    }),
-  );
-
-  return found;
 }
 
 export async function getNxDotnetProjects(

--- a/project.json
+++ b/project.json
@@ -8,6 +8,5 @@
         "storage": "build/local-registry/storage"
       }
     }
-  },
-  "name": "."
+  }
 }

